### ISSUE TITLE
USB Serial Issues

### ISSUE THREAD

**Issue by IoTPanic** [RAKWireless/WisBlock#11]
Hope all is well,

I have noticed strange behavior when using USB serial with the wisblocks, if I don't send a serial message in quick enough intervals from the wisblock to my machine, the connection drops, then eventually there is a reset. [I have the following video showing a basic demo to show the specific issue](https://qubitro-my.sharepoint.com/:v:/p/samuel_archibald/EcAcyrh2fPtOiXsMK3A3ZRYBRyeYI48cA8E9vJKh8KcHIw?e=wZ9xOJ). If I dont have USB connected and have it break out of the `while(!Serial)` after a few seconds to run on a battery, there is no issue and it never resets even if it prints to serial during operation.

I have noticed this issue since I started using the wisblocks in the past week, but I was able to get around the issue by printing a UTF-8 blank character every few hundred milliseconds. I noticed that if I send serial data to the device, [it seems like a buffer on the device side empties as shown in this video.](https://qubitro-my.sharepoint.com/:v:/p/samuel_archibald/ERr5j070FLNMt9o0ov-TRdYBM8XqgKSCsSwfJu_dkQiRfg?e=Piegcj) If I need to update something, let me know, I have not noticed this issue with any other device on my desktop.

## Setup
* System (`uname -a`) - `Linux manjaro-tank 5.10.23-1-MANJARO`
* Framework - PlatformIO w/ Arduino using the getting started guide in this repo

## Demo Code
```c++
#include <Arduino.h>

// Value of 300ms is OK, 3 or 30 seconds not so much
#define INTERVAL 3000

unsigned long next = 0;
unsigned i = 0;

void setup() {
  pinMode(LED_BLUE, OUTPUT);
  digitalWrite(LED_BLUE, HIGH);
  Serial.begin(9600);
  while(!Serial) {
    // Need otherwise serial port is lost if I dont get to it fast enough
    Serial.println(); 
  }
  Serial.println("Hello! I'm starting now");
  digitalWrite(LED_BLUE, LOW);
}

void loop() {
  // Delays also have the same effect, I wanted any possible hidden timer or rtos setup to not be hindered
  if(millis()>=next) {
    digitalWrite(LED_BLUE, HIGH);
    Serial.print("Printing to usb serial ");
    Serial.println(i++);
    // So I can see the led with my human eyeballs
    delay(10); 
    digitalWrite(LED_BLUE, LOW);
    next = INTERVAL + millis();
  } 
}
```

**Comment by beegee-tokyo** [comment]
I am using WisBlock now for nearly a year and I never saw what you are reporting. I will ask RAK R&D to have a look into it.

**Comment by jpmeijers** [comment]
This issues sounds very similar to the behaviour that I have seen.

**Comment by beegee-tokyo** [comment]
Our HW team says they cannot see the problem.
I tried as well, using exactly your code, but set the time between two Serial.print() to 10 seconds. Compiled on PlatformIO.
Started test at 16:43:55
![image](https://user-images.githubusercontent.com/512690/114372321-04178180-9bb4-11eb-8917-f473366c1d70.png)
and let it run for nearly two hours without any problem
![image](https://user-images.githubusercontent.com/512690/114380777-07fbd180-9bbd-11eb-9d40-08d1ecfac4a1.png)

I tested on Windows 10 (no MAC around here). Not sure if that makes a difference.

**Comment by jpmeijers** [comment]
Ubuntu 20.04 here, using Arduino 1.8.13.

I've tried reproducing this issue today, but so far it keeps on working. I wonder if the bootloader upgrade didn't solve this issue for me.

**Comment by IoTPanic** [comment]
I went ahead and installed a Windows 10 VM in virtualbox and passed the USB device to the guest, works just fine there, while having the same issues in the linux host. @jpmeijers can you try with 10 second delay? If that happens to work fine could I have the output of `uname -a` and  `sudo cat /sys/kernel/debug/usb/devices`?

**Comment by beegee-tokyo** [comment]
Closing at it is not related to WisBlock

**Comment by IoTPanic** [comment]
I disagree, other nRF52 systems do not experience these issues with my machine, the issue is specific to WisBlocks, we seem to be narrowing down the issue to be related to the host OS. If other people who also use nRF52 boards with linux then use WisBlocks, this will be the place they look to talk about the issue.

**Comment by bhutch29** [comment]
I know it's been years but I am having this exact problem. Thought I was going a bit crazy until I managed to find this, so thank you for submitting it.

Any chance anyone has identified a solution, other than frequently printing to keep things moving?

Other devices work on all operating systems, Wisblock NRF52 devices work on other operating systems, but Wisblock NRF52 devices don't work on Linux. Any delay more than about 1 second causes the Serial data to completely stop. Sending data from either end temporarily fixes things. Tried different serial monitors, different baud rates, nothing fixes it.
